File: algorithms/root_finding_algorithms.cpp
#include <iostream>
#include <cmath>
#include <functional>

// Newton-Raphson solver
double newton_solver(const std::function<double(double)>& func, const std::function<double(double)>& derivative, double x0, double epsilon = 1e-6, int max_iterations = 100) {
    for (int i = 0; i < max_iterations; ++i) {
        double f_x0 = func(x0);
        double f_prime_x0 = derivative(x0);

        double x1 = x0 - f_x0 / f_prime_x0;

        if (std::abs(x1 - x0) < epsilon) {
            return x1; // Converged to a root
        }

        x0 = x1;
    }

    throw std::runtime_error("Newton-Raphson method did not converge");
}

// Secant solver
double secant_solver(const std::function<double(double)>& func, double x0, double x1, double epsilon = 1e-6, int max_iterations = 100) {
    for (int i = 0; i < max_iterations; ++i) {
        double f_x0 = func(x0);
        double f_x1 = func(x1);

        if (std::abs(f_x1) < epsilon) {
            return x1;
        }

        double x_new = x1 - f_x1 * (x1 - x0) / (f_x1 - f_x0);

        x0 = x1;
        x1 = x_new;
    }

    throw std::runtime_error("Secant method did not converge");
}

// Bisection solver
double bisection_solver(const std::function<double(double)>& func, double x0, double x1, double epsilon = 1e-6, int max_iterations = 100) {
    for (int i = 0; i < max_iterations; ++i) {
        double mid = (x0 + x1) / 2;
        double f_mid = func(mid);

        if (std::abs(f_mid) < epsilon) {
            return mid;
        }

        if (f_mid < 0) {
            x0 = mid;
        } else {
            x1 = mid;
        }
    }

    throw std::runtime_error("Bisection method did not converge");
}

int main() {
    // Example
    auto func = [](double x) { return x * x - 4 * x - 7; };
    auto derivative = [](double x) { return 2 * x - 4; };

    double solution = newton_solver(func, derivative, 5);
    std::cout << "Newton-Raphson solution: " << solution << std::endl;

    solution = secant_solver(func, 5, 10);
    std::cout << "Secant solution: " << solution << std::endl;

    solution = bisection_solver(func, 5, 10);
    std::cout << "Bisection solution: " << solution << std::endl;

    return 0;
}

File: algorithms/root_finding_algorithms.py
import sympy as sp


def newton_solver(expression, x0, epsilon=1e-6, max_iterations=100, variable='x'):
    try:
        expr = sp.sympify(expression)

    except Exception as e:
        raise ValueError(f"Error: {str(e)}")

    x = sp.symbols(variable)  # differentiate function with respect to { arg }
    derivative = sp.diff(expr, x)

    for i in range(max_iterations):

        x1 = x0 - expr.subs(x, x0) / derivative.subs(x, x0)

        if abs(x1 - x0) < epsilon:
            return float(x1)  # Converged to a root
        x0 = x1

    raise ValueError("Newton-Raphson method did not converge")


def secant_solver(expression, x0, x1, epsilon=1e-6, max_iterations=100, variable='x', override=False):
    try:
        expr = sp.sympify(expression)
        x = sp.symbols(variable)  # differentiate function with respect to { arg }

        if expr.subs(x, x0) * expr.subs(x, x1) >= 0 and not override:
            raise ValueError(
                "Initial values x0 and x1 should have different signs for f(x). Update x0 or x1, or set override arg equal to True (not recommended).")

    except Exception as e:
        raise ValueError(f"Error: {str(e)}")

    for i in range(max_iterations):

        f_x0 = expr.subs(x, x0)
        f_x1 = expr.subs(x, x1)

        if abs(f_x1) < epsilon:
            return float(x1)

        x_new = x1 - f_x1 * (x1 - x0) / (f_x1 - f_x0)

        x0 = x1
        x1 = x_new

    raise ValueError("Secant method did not converge")


def bisection_solver(expression, x0, x1, epsilon=1e-6, max_iterations=100, variable='x', override=False):
    try:
        expr = sp.sympify(expression)
        x = sp.symbols(variable)  # differentiate function with respect to { arg }

        if expr.subs(x, x0) * expr.subs(x, x1) >= 0 and not override:
            raise ValueError(
                "Initial values x0 and x1 should have different signs for f(x). Update x0 or x1, or set override arg equal to True (not recommended).")

    except Exception as e:
        raise ValueError(f"Error: {str(e)}")

    for i in range(max_iterations):

        mid = (x0 + x1) / 2
        f_mid = expr.subs(x, mid)

        if abs(f_mid) < epsilon:
            return float(mid)

        if f_mid < 0:
            x0 = mid
        else:
            x1 = mid

    raise ValueError("Secant method did not converge")


if __name__ == '__main__':

    # Example
    expression_string = "x**2 - 4*x - 7"

    solution = newton_solver(expression=expression_string, x0=5)
    print(solution)

    solution = secant_solver(expression=expression_string, x0=5, x1=10)
    print(solution)

    solution = bisection_solver(expression=expression_string, x0=5, x1=10)
    print(solution)
